# Restrict types so that only members of the same kind are allowed in them (`functional/no-mixed-types`)

üíº This rule is enabled in the following configs: ‚òëÔ∏è `lite`, `no-other-paradigms`, ‚úÖ `recommended`, üîí `strict`.

<!-- end auto-generated rule header -->

This rule enforces that an aliased type literal or an interface only has one type of members, eg. only data properties or only functions.

## Rule Details

Mixing functions and data properties in the same type is a sign of object-orientation style.

### ‚ùå Incorrect

<!-- eslint-skip -->

```ts
/* eslint functional/no-mixed-types: "error" */

type Foo = {
  prop1: string;
  prop2: () => string;
};
```

### ‚úÖ Correct

```ts
/* eslint functional/no-mixed-types: "error" */

type Foo = {
  prop1: string;
  prop2: number;
};
```

```ts
/* eslint functional/no-mixed-types: "error" */

type Foo = {
  prop1: () => string;
  prop2: () => () => number;
};
```

## Limitations

This rule will only check alias type literal declarations and interface declarations. Advanced types will not be checked.
For example union and intersection types will not be checked.

## Options

This rule accepts an options object of the following type:

```ts
type Options = {
  checkInterfaces: boolean;
  checkTypeLiterals: boolean;
};
```

### Default Options

```ts
const defaults = {
  checkInterfaces: true,
  checkTypeLiterals: true,
};
```

### checkInterfaces

If true, interface declarations will be checked.

### checkTypeLiterals

If true, aliased type literal declarations will be checked.
