# Disallow this access (`functional/no-this-expressions`)

ğŸ’¼ğŸš« This rule is enabled in the following configs: `no-other-paradigms`, ğŸ”’ `strict`. This rule is _disabled_ in the following configs: â˜‘ï¸ `lite`, âœ… `recommended`.

<!-- end auto-generated rule header -->

## Rule Details

This rule is a companion rule to the [no-classes](./no-classes.md) rule.
See the its docs for more info.

### âŒ Incorrect

<!-- eslint-skip -->

```js
/* eslint functional/no-this-expressions: "error" */

const foo = this.value + 17;
```

### âœ… Correct

```js
/* eslint functional/no-this-expressions: "error" */

const foo = object.value + 17;
```
